src/thermo/ThermoPhase.cpp

```diff
@@ -689,6 +689,617 @@ void ThermoPhase::setState_SPorSV(double Starget, double p,
         throw CanteraError("ThermoPhase::setState_SPorSV (SP)", ErrString);
     }
 }
+double ThermoPhase::getStoichAirFuelRatio_X(const compositionMap& fuelComp,
+                                            const compositionMap& oxComp) const
+{
+    vector_fp Xf(getCompositionFromMap(fuelComp));
+    vector_fp Xo(getCompositionFromMap(oxComp));
+    return getStoichAirFuelRatio_X(Xf.data(), Xo.data());
+}
+
+double ThermoPhase::getStoichAirFuelRatio_X(const std::string& fuelComp,
+                                            const std::string& oxComp) const
+{
+    return getStoichAirFuelRatio_X(
+            parseCompString(fuelComp.find(":") != std::string::npos ? fuelComp : fuelComp+":1.0"),
+            parseCompString(oxComp.find(":") != std::string::npos ? oxComp : oxComp+":1.0"));
+}
+
+double ThermoPhase::getStoichAirFuelRatio_X(const vector_fp& fuelComp,
+                                            const vector_fp& oxComp) const
+{
+    return getStoichAirFuelRatio_X(fuelComp.data(), oxComp.data());
+}
+
+double ThermoPhase::getStoichAirFuelRatio_X(const double* fuelComp,
+                                            const double* oxComp) const
+{
+    // indices of elements
+    auto iC = elementIndex("C");
+    auto iS = elementIndex("S");
+    auto iO = elementIndex("O");
+    auto iH = elementIndex("H");
+
+    // non-normalized element mole fractions
+    auto Co = 0.0, Cf = 0.0;
+    auto So = 0.0, Sf = 0.0;
+    auto Oo = 0.0, Of = 0.0;
+    auto Ho = 0.0, Hf = 0.0;
+    auto sum_Xf = 0.0, sum_Xo = 0.0;
+
+    for (size_t i=0; i!=m_kk; ++i)
+    {
```

Comment:
A couple of style notes applicable here and elsewhere:
```suggestion
    for (size_t i = 0; i != m_kk; ++i) {
```
The opening brace should be on the same line for `if` statements as well.